### DISCUSSION TITLE
Send large data to client

### DISCUSSION THREAD

**Discussion by falkoschindler** [elimintz/justpy#274] [Q&A] [answered]
I'm trying to send a number of images to the client side of a custom component. Images update frequently and the application runs over a remote connection, so bandwidth is an issue.

My first approach was to add the raw JPEG bytes into the component's options dictionary. But each time other options change, the whole dictionary is sent to the client, causing unnecessary traffic.

I experimented with removing the image data once they arrived, but I still can't manage to avoid multiple transmissions. It should be possible to block updates as long as the client did not acknowledge the image transmission, but I already struggled for hours and the resulting implementation appears quite complex for such rather simple requirement.

My next idea is to only transmit an ID or URL and let the client get the image via HTTP. But how can I define an image endpoint with JustPy? As far as I can tell, a jp.Route has to return a WebPage.

Maybe I'm missing something completely obvious. But the documentation on https://justpy.io/tutorial/advanced_components/ only provides little insight into what is possible and how one is thought to approach things.

**Comment by elimintz** [comment, marked as answer]
Could you please elaborate why you would need a special component in this case? What is the functionality you need that only an advance component could provide? 

As for sending endpoints for images, please take a look at this example: https://justpy.io/tutorial/request_object/#dog-example

While the request returns a webpage, you could use the request object to specify the src of images on a page. You could also use an event handler to add images to a page or change the src attribute of an image already on the page.

**Reply by falkoschindler** [reply]
We are working on an application with a 3d view of a camera setup. It is based on [Three.js](http://threejs.org/) and uses texture mapping to show projected images. The texture loader can either use raw image data or URLs, but not images from the DOM. Our custom component is already working, but had above-mentioned problems.

Currently our solution is to introduce a new Starlette endpoint (inserted at index 0 to be evaluated first) that provides image data for a given image id. The browser cache takes care of loading each image only once, even after refreshing the page.

```
def get_image_data(request, **kwargs):

    data = image_data[request.path_params['id']]
    return starlette.responses.Response(content=data, media_type='image/jpeg')

app.routes.insert(0, starlette.routing.Route('/imagedata/{id}', get_image_data))
```

So basically we solved this issue. But maybe there is an easier way to address this requirement in JustPy. I was thinking about something like a method to send an arbitrary payload from Python to JavaScript, independent of the automatic component updates.

**Reply by elimintz** [reply]
I see what you mean. In the upcoming version, a JustPy request handler can return either a  starlette Response or an object that has WebPage as a subclass. If the returned object is of type Response, it is just returned to webpage without further processing. In the next version you could do something very similar as above but use the JustPy decorator for routing. But what you did is great. The fact that a JustPy app is a starlette app, provides a lot of flexibility. 

Please also take a look at https://github.com/elimintz/justpy/issues/180
